jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -495,16 +510,33 @@ public MethodHandle getBoundMethodHandle() {
 			MethodType nativeMethodType = methodType(Object.class, Addressable.class, SegmentAllocator.class, long[].class);
 			/*[ENDIF] JAVA_SPEC_VERSION >= 21 */
 
-			MethodHandle boundHandle = lookup.bind(this, "runNativeMethod", nativeMethodType);
+			boundMH = lookup.bind(this, "runNativeMethod", nativeMethodType);
 
 			/* Replace the original handle with the specified types of the C function. */
-			boundHandle = permuteMH(boundHandle, funcMethodType);
-			return boundHandle;
+			boundMH = permuteMH(boundMH, funcMethodType);
+
+			/*[IF JAVA_SPEC_VERSION >= 22]*/
+			return isFfiProtoOn ? getNativeMHWithInvokeCache(boundMH) : boundMH;
+			/*[ELSE] JAVA_SPEC_VERSION >= 22 */
+			return boundMH;
+			/*[ENDIF] JAVA_SPEC_VERSION >= 22 */
 		} catch (ReflectiveOperationException e) {
 			throw new InternalError(e);
 		}
 	}
 
+	/*[IF JAVA_SPEC_VERSION >= 22]*/
+	/* Generate the native MethodHandle with the resolved bound handle. */
+	private MethodHandle getNativeMHWithInvokeCache(MethodHandle handle) {
+		nativeMH = SharedSecrets.getJavaLangInvokeAccess().nativeMethodHandle(handle);
+		/* The cache array declared in the NativeMethodHandle (OpenJDK) stores MemberName
+		 * and appendix resolved from the bounded handle.
+		 */
+		resolveBoundHandle();
```

Comment:
I think you should explicitly pass the arg to `resolveBoundHandle(nativeMH)` and not rely on the existing stack shape to always be as it is. Its safer that way.

Reply:
I used to think about that but I went straightforward to the current solution with the macros to access `nativeMH` later to ensure the `nativeMH` object was always the latest one even though it was updated by GC. But I can try it out again to see whether `resolveBoundHandle(nativeMH)` works in our case. If it still crash due to the outdated `nativeMH` by GC, we have to go with the current implementation.

Reply:
Updated the code to pass `nativeMH` as argument to the native code via `resolveBoundHandle(nativeMH)` which must be created as a local reference to ensure the latest `nativeMH` object (updated by GC) is selected in resolution.